api/server/runner.go

```diff
@@ -55,14 +55,16 @@ var (
 	bufPool = &sync.Pool{New: func() interface{} { return new(bytes.Buffer) }}
 )
 
-// TODO it would be nice if we could make this have nothing to do with the gin.Context but meh
-// TODO make async store an *http.Request? would be sexy until we have different api format...
-func (s *Server) serve(c *gin.Context, app *models.App, path string) error {
+// CallFunction exposed to become the API extension, to let API listeners to be capable to call a functions
+// defined by app and path
+func (s *Server) CallFunction(app *models.App, path string, req *http.Request,
+	respWriter http.ResponseWriter) (*string, io.Reader, error) {
```

Comment:
with the general idea here, thought we had done this already for some reason. this does seem an odd signature for an extension to be able to fulfill. most of our extensions are either based around http req/resp or the app/route and don't seem likely to have a handle on all of these. a lot of the stuff in here is also pretty specific to our http endpoint and not necessarily relevant to the extension caller (error handling, wait time, async, etc).

seems like:

```
var out bytes.Buffer
call, err := Server.Agent.GetCall(agent.FromRequest(req), agent.WithWriter(&out))
err := Server.Agent.Submit(call)
```

is more in line with what extensions would want, and it's flexible for the extension to do this based on whatever they have, be it a request or an app/route (we can add options to make this easier), or build a model, etc. we could tidy this up into a one liner easily enough. the proposed plumbing here just seems a little off. exposing agent is easy, not sure if it's the right plumbing, either, though.

Reply:
Agreed, haven't thought about exposing the agent directly. So, yeah, i assume expose agent suppose to work fine for calling the functions directly.